File: 417-pacific-atlantic-water-flow/pacific-atlantic-water-flow.cpp
class Solution {
public:
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        int m = heights.size(), n = heights[0].size();
        vector<vector<int>> res;
        vector<vector<bool>> pac(m, vector<bool>(n, false));
        vector<vector<bool>> atl(m, vector<bool>(n, false));
        deque<pair<int, int>> pq, aq;

        for (int i = 0; i < m; i++) {
            pq.push_back({i, 0});
            aq.push_back({i, n - 1});
            pac[i][0] = true;
            atl[i][n - 1] = true;
        }
        for (int j = 0; j < n; j++) {
            pq.push_back({0, j});
            aq.push_back({m - 1, j});
            pac[0][j] = true;
            atl[m - 1][j] = true;
        }

        int dirs[5] = {0, 1, 0, -1, 0};
        auto bfs = [&](deque<pair<int, int>>& q, vector<vector<bool>>& ocean) {
            while (!q.empty()) {
                auto [x, y] = q.front();
                q.pop_front();
                for (int d = 0; d < 4; d++) {
                    int nx = x + dirs[d], ny = y + dirs[d + 1];
                    if (nx >= 0 && ny >= 0 && nx < m && ny < n &&
                        !ocean[nx][ny] && heights[nx][ny] >= heights[x][y]) {
                        ocean[nx][ny] = true;
                        q.push_back({nx, ny});
                    }
                }
            }
        };

        bfs(pq, pac);
        bfs(aq, atl);

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (pac[i][j] && atl[i][j])
                    res.push_back({i, j});
            }
        }
        return res;
    }
};

File: 417-pacific-atlantic-water-flow/pacific-atlantic-water-flow.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        m, n = len(heights), len(heights[0])
        pac, atl= set(), set(),
        pq, aq = deque([]), deque([])
        for i in range(m):
            pac.add((i,0))
            pq.append((i,0))
            
            atl.add((i,n-1))
            aq.append((i,n-1))
            
        for j in range(n):
            pac.add((0,j))
            pq.append((0,j))
            
            atl.add((m-1,j))
            aq.append((m-1,j))
            
        dirs = [(0,1),(0,-1),(1,0),(-1,0)]
        while pq:
            x,y = pq.popleft()
            for dx,dy in dirs:
                nx,ny = x+dx, y+dy
                if nx>=0 and nx<m and ny>=0 and ny<n and (nx,ny) not in pac and heights[nx][ny] >= heights[x][y]:
                    pac.add((nx,ny))
                    pq.append((nx,ny))
        while aq:
            x,y = aq.popleft()
            for dx,dy in dirs:
                nx,ny = x+dx, y+dy
                if nx>=0 and nx<m and ny>=0 and ny<n and (nx,ny) not in atl and heights[nx][ny] >= heights[x][y]:
                    atl.add((nx,ny))
                    aq.append((nx,ny))
        res = []
        for loc in pac:
            if loc in atl:
                res.append(loc)
        return res          